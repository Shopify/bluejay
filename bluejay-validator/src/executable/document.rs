mod error;
mod orchestrator;
mod path;
mod rule;
pub mod rules;
mod variable_definition_input_type;
mod visitor;

pub use error::{ArgumentError, DirectiveError, Error};
pub use orchestrator::{BuiltinRulesValidator, Orchestrator};
pub use path::{Path, PathRoot};
pub use rule::Rule;
pub use rules::BuiltinRules;
pub use variable_definition_input_type::VariableDefinitionInputType;
pub use visitor::Visitor;
