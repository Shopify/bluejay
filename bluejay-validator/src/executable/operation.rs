mod analyzer;
pub mod analyzers;
mod orchestrator;
mod variable_values;
mod visitor;

pub use analyzer::Analyzer;
pub use orchestrator::Orchestrator;
pub use variable_values::{OperationDefinitionValueEvaluationExt, VariableValues};
pub use visitor::Visitor;
