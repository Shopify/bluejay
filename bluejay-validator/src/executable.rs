mod cache;
mod error;
mod rule;
pub mod rules;
mod validator;
mod variable_definition_input_type_reference;
mod visitor;

pub use cache::Cache;
pub use error::{ArgumentError, DirectiveError, Error};
pub use rule::Rule;
use rules::Rules;
pub use validator::{RulesValidator, Validator};
use variable_definition_input_type_reference::VariableDefinitionInputTypeReference;
use visitor::Visitor;
