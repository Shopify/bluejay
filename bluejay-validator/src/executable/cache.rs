use crate::executable::VariableDefinitionInputType;
use bluejay_core::definition::SchemaDefinition;
use bluejay_core::executable::{
    ExecutableDocument, FragmentDefinition, OperationDefinition, VariableDefinition,
};
use bluejay_core::AsIter;
use std::collections::HashMap;

pub struct Cache<'a, E: ExecutableDocument, S: SchemaDefinition> {
    variable_definition_input_type_references:
        HashMap<&'a E::VariableType, VariableDefinitionInputType<'a, S::BaseInputType>>,
    indexed_fragment_definitions: HashMap<&'a str, &'a E::FragmentDefinition>,
}

impl<'a, E: ExecutableDocument, S: SchemaDefinition> Cache<'a, E, S> {
    pub fn new(executable_document: &'a E, schema_definition: &'a S) -> Self {
        let variable_definition_input_type_references =
            HashMap::from_iter(executable_document.operation_definitions().iter().flat_map(
                |operation_definition: &'a E::OperationDefinition| {
                    let variable_definitions_iterator = operation_definition
                        .as_ref()
                        .variable_definitions()
                        .map(|variable_definitions: &'a E::VariableDefinitions| -> <E::VariableDefinitions as AsIter>::Iterator<'a> {
                            variable_definitions.iter()
                        });

                    variable_definitions_iterator
                        .into_iter()
                        .flatten()
                        .filter_map(|variable_definition| {
                            let type_reference = variable_definition.r#type();
                            VariableDefinitionInputType::try_from((
                                schema_definition,
                                type_reference,
                            ))
                            .ok()
                            .map(|vditr| (type_reference, vditr))
                        })
                },
            ));
        let indexed_fragment_definitions = HashMap::from_iter(
            executable_document
                .fragment_definitions()
                .as_ref()
                .iter()
                .map(|fragment_definition| (fragment_definition.name(), fragment_definition)),
        );
        Self {
            variable_definition_input_type_references,
            indexed_fragment_definitions,
        }
    }

    pub fn variable_definition_input_type_reference(
        &self,
        variable_type: &E::VariableType,
    ) -> Option<&VariableDefinitionInputType<'a, S::BaseInputType>> {
        self.variable_definition_input_type_references
            .get(variable_type)
    }

    pub fn fragment_definition(&self, name: &str) -> Option<&'a E::FragmentDefinition> {
        self.indexed_fragment_definitions.get(name).copied()
    }
}
