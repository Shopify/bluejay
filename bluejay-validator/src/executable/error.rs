use bluejay_core::definition::{
    DirectiveDefinition, FieldDefinition, InputValueDefinition, SchemaDefinition,
    TypeDefinitionReferenceFromAbstract,
};
use bluejay_core::executable::{ExecutableDocument, OperationDefinitionFromExecutableDocument};
use bluejay_core::{
    call_const_wrapper_method, ArgumentWrapper, Directive, DirectiveWrapper, OperationType,
};
#[cfg(feature = "parser-integration")]
use bluejay_parser::{
    ast::executable::ExecutableDocument as ParserExecutableDocument,
    error::{Annotation, Error as ParserError},
    HasSpan,
};
use itertools::join;

pub enum Error<'a, E: ExecutableDocument, S: SchemaDefinition> {
    NonUniqueOperationNames {
        name: &'a str,
        operations: Vec<&'a E::ExplicitOperationDefinition>,
    },
    NotLoneAnonymousOperation {
        anonymous_operations: Vec<&'a OperationDefinitionFromExecutableDocument<E>>,
    },
    SubscriptionRootNotSingleField {
        operation: &'a OperationDefinitionFromExecutableDocument<E>,
    },
    FieldDoesNotExistOnType {
        field: &'a E::Field,
        r#type: &'a TypeDefinitionReferenceFromAbstract<S::TypeDefinitionReference>,
    },
    OperationTypeNotDefined {
        operation: &'a E::ExplicitOperationDefinition,
    },
    LeafFieldSelectionNotEmpty {
        selection_set: &'a E::SelectionSet,
        r#type: &'a S::OutputTypeReference,
    },
    NonLeafFieldSelectionEmpty {
        field: &'a E::Field,
        r#type: &'a S::OutputTypeReference,
    },
    ArgumentDoesNotExistOnField {
        argument: &'a E::Argument<false>,
        field_definition: &'a S::FieldDefinition,
    },
    ArgumentDoesNotExistOnDirective {
        argument: ArgumentWrapper<'a, E::Argument<true>, E::Argument<false>>,
        directive_definition: &'a S::DirectiveDefinition,
    },
    NonUniqueArgumentNames {
        arguments: Vec<ArgumentWrapper<'a, E::Argument<true>, E::Argument<false>>>,
        name: &'a str,
    },
    FieldMissingRequiredArguments {
        field: &'a E::Field,
        field_definition: &'a S::FieldDefinition,
        missing_argument_definitions: Vec<&'a S::InputValueDefinition>,
        arguments_with_null_values: Vec<&'a E::Argument<false>>,
    },
    DirectiveMissingRequiredArguments {
        directive: DirectiveWrapper<'a, E::Directive<true>, E::Directive<false>>,
        directive_definition: &'a S::DirectiveDefinition,
        missing_argument_definitions: Vec<&'a S::InputValueDefinition>,
        arguments_with_null_values: Vec<ArgumentWrapper<'a, E::Argument<true>, E::Argument<false>>>,
    },
    NonUniqueFragmentDefinitionNames {
        name: &'a str,
        fragment_definitions: Vec<&'a E::FragmentDefinition>,
    },
    FragmentDefinitionTargetTypeDoesNotExist {
        fragment_definition: &'a E::FragmentDefinition,
    },
    InlineFragmentTargetTypeDoesNotExist {
        inline_fragment: &'a E::InlineFragment,
    },
    FragmentDefinitionTargetTypeNotComposite {
        fragment_definition: &'a E::FragmentDefinition,
    },
    InlineFragmentTargetTypeNotComposite {
        inline_fragment: &'a E::InlineFragment,
    },
    FragmentDefinitionUnused {
        fragment_definition: &'a E::FragmentDefinition,
    },
    FragmentSpreadTargetUndefined {
        fragment_spread: &'a E::FragmentSpread,
    },
    FragmentSpreadCycle {
        fragment_definition: &'a E::FragmentDefinition,
        fragment_spread: &'a E::FragmentSpread,
    },
    FieldSelectionsDoNotMergeIncompatibleTypes {
        selection_set: &'a E::SelectionSet,
        field_a: &'a E::Field,
        field_definition_a: &'a S::FieldDefinition,
        field_b: &'a E::Field,
        field_definition_b: &'a S::FieldDefinition,
    },
    FieldSelectionsDoNotMergeDifferingNames {
        selection_set: &'a E::SelectionSet,
        field_a: &'a E::Field,
        field_b: &'a E::Field,
    },
    FieldSelectionsDoNotMergeDifferingArguments {
        selection_set: &'a E::SelectionSet,
        field_a: &'a E::Field,
        field_b: &'a E::Field,
    },
    FragmentSpreadIsNotPossible {
        fragment_spread: &'a E::FragmentSpread,
        parent_type: &'a TypeDefinitionReferenceFromAbstract<S::TypeDefinitionReference>,
    },
    InlineFragmentSpreadIsNotPossible {
        inline_fragment: &'a E::InlineFragment,
        parent_type: &'a TypeDefinitionReferenceFromAbstract<S::TypeDefinitionReference>,
    },
}

#[cfg(feature = "parser-integration")]
impl<'a, S: SchemaDefinition> From<Error<'a, ParserExecutableDocument<'a>, S>> for ParserError {
    fn from(value: Error<'a, ParserExecutableDocument<'a>, S>) -> Self {
        match value {
            Error::NonUniqueOperationNames { name, operations } => Self::new(
                format!("Multiple operation definitions named `{name}`"),
                None,
                operations
                    .iter()
                    .filter_map(|operation| {
                        operation.name().map(|operation_name| {
                            Annotation::new(
                                format!("Operation definition with name `{name}`"),
                                operation_name.span().clone(),
                            )
                        })
                    })
                    .collect(),
            ),
            Error::NotLoneAnonymousOperation {
                anonymous_operations,
            } => Self::new(
                "Anonymous operation not lone operation in document",
                None,
                anonymous_operations
                    .iter()
                    .map(|operation| {
                        Annotation::new(
                            "Anonymous operation definition",
                            operation.selection_set().span().clone(),
                        )
                    })
                    .collect(),
            ),
            Error::SubscriptionRootNotSingleField { operation } => Self::new(
                "Subscription root is not a single field",
                Some(Annotation::new(
                    "Selection set contains multiple fields",
                    operation.selection_set().span().clone(),
                )),
                Vec::new(),
            ),
            Error::FieldDoesNotExistOnType { field, r#type } => Self::new(
                format!(
                    "Field `{}` does not exist on type `{}`",
                    field.name().as_ref(),
                    r#type.name()
                ),
                Some(Annotation::new(
                    format!("Field does not exist on type `{}`", r#type.name()),
                    field.name().span().clone(),
                )),
                Vec::new(),
            ),
            Error::OperationTypeNotDefined { operation } => Self::new(
                format!(
                    "Schema does not define a {} root",
                    OperationType::from(operation.operation_type()),
                ),
                Some(Annotation::new(
                    format!(
                        "Schema does not define a {} root",
                        OperationType::from(operation.operation_type()),
                    ),
                    operation.operation_type().span().clone(),
                )),
                Vec::new(),
            ),
            Error::LeafFieldSelectionNotEmpty {
                selection_set,
                r#type,
            } => Self::new(
                format!(
                    "Selection on field of leaf type `{}` was not empty",
                    r#type.as_ref().display_name()
                ),
                Some(Annotation::new(
                    "Selection set on field of leaf type must be empty",
                    selection_set.span().clone(),
                )),
                Vec::new(),
            ),
            Error::NonLeafFieldSelectionEmpty { field, r#type } => Self::new(
                format!(
                    "No selection on field of non-leaf type `{}`",
                    r#type.as_ref().display_name()
                ),
                Some(Annotation::new(
                    "Fields of non-leaf types must have a selection",
                    field.name().span().clone(),
                )),
                Vec::new(),
            ),
            Error::ArgumentDoesNotExistOnField {
                argument,
                field_definition,
            } => Self::new(
                format!(
                    "Field `{}` does not define an argument named `{}`",
                    field_definition.name(),
                    argument.name().as_ref(),
                ),
                Some(Annotation::new(
                    "No argument definition with this name",
                    argument.name().span().clone(),
                )),
                Vec::new(),
            ),
            Error::ArgumentDoesNotExistOnDirective {
                argument,
                directive_definition,
            } => {
                let name = call_const_wrapper_method!(ArgumentWrapper, argument, name);
                Self::new(
                    format!(
                        "Directive `{}` does not define an argument named `{}`",
                        directive_definition.name(),
                        name.as_ref(),
                    ),
                    Some(Annotation::new(
                        "No argument definition with this name",
                        name.span().clone(),
                    )),
                    Vec::new(),
                )
            }
            Error::NonUniqueArgumentNames { arguments, name } => Self::new(
                format!("Multiple arguments with name `{name}`"),
                None,
                arguments
                    .into_iter()
                    .map(|argument| {
                        Annotation::new(
                            format!("Argument with name `{name}`"),
                            call_const_wrapper_method!(ArgumentWrapper, argument, name)
                                .span()
                                .clone(),
                        )
                    })
                    .collect(),
            ),
            Error::FieldMissingRequiredArguments {
                field,
                field_definition: _,
                missing_argument_definitions,
                arguments_with_null_values,
            } => {
                let missing_argument_names = join(
                    missing_argument_definitions
                        .into_iter()
                        .map(InputValueDefinition::name),
                    ", ",
                );
                let span = match field.arguments() {
                    Some(arguments) => field.name().span().merge(arguments.span()),
                    None => field.name().span().clone(),
                };
                Self::new(
                    format!(
                        "Field `{}` missing argument(s): {missing_argument_names}",
                        field.response_key()
                    ),
                    Some(Annotation::new(
                        format!("Missing argument(s): {missing_argument_names}"),
                        span,
                    )),
                    arguments_with_null_values
                        .into_iter()
                        .map(|argument| {
                            Annotation::new(
                                "`null` value provided for required argument",
                                argument.span().clone(),
                            )
                        })
                        .collect(),
                )
            }
            Error::DirectiveMissingRequiredArguments {
                directive,
                directive_definition: _,
                missing_argument_definitions,
                arguments_with_null_values,
            } => {
                let missing_argument_names = join(
                    missing_argument_definitions
                        .into_iter()
                        .map(InputValueDefinition::name),
                    ", ",
                );
                let directive_name = call_const_wrapper_method!(DirectiveWrapper, directive, name);
                let span = call_const_wrapper_method!(DirectiveWrapper, directive, span);
                Self::new(
                    format!(
                        "Directive `{directive_name}` missing argument(s): {missing_argument_names}",
                    ),
                    Some(Annotation::new(
                        format!("Missing argument(s): {missing_argument_names}"),
                        span.clone(),
                    )),
                    arguments_with_null_values
                        .into_iter()
                        .map(|argument| Annotation::new(
                    "`null` value provided for required argument",
                        call_const_wrapper_method!(ArgumentWrapper, argument, span).clone(),
                        ))
                        .collect(),
                    )
            }
            Error::NonUniqueFragmentDefinitionNames {
                name,
                fragment_definitions,
            } => Self::new(
                format!("Multiple fragment definitions named `{name}`"),
                None,
                fragment_definitions
                    .iter()
                    .map(|fragment_definition| {
                        Annotation::new(
                            format!("Fragment definition with name `{name}`"),
                            fragment_definition.name().span().clone(),
                        )
                    })
                    .collect(),
            ),
            Error::FragmentDefinitionTargetTypeDoesNotExist {
                fragment_definition,
            } => Self::new(
                format!(
                    "No type definition with name `{}`",
                    fragment_definition.type_condition().named_type().as_ref()
                ),
                Some(Annotation::new(
                    "No type with this name",
                    fragment_definition
                        .type_condition()
                        .named_type()
                        .span()
                        .clone(),
                )),
                Vec::new(),
            ),
            Error::InlineFragmentTargetTypeDoesNotExist { inline_fragment } => Self::new(
                format!(
                    "No type definition with name `{}`",
                    inline_fragment
                        .type_condition()
                        .map(|tc| tc.named_type().as_ref())
                        .unwrap_or_default()
                ),
                inline_fragment.type_condition().map(|tc| {
                    Annotation::new("No type with this name", tc.named_type().span().clone())
                }),
                Vec::new(),
            ),
            Error::FragmentDefinitionTargetTypeNotComposite {
                fragment_definition,
            } => Self::new(
                format!(
                    "`{}` is not a composite type",
                    fragment_definition.type_condition().named_type().as_ref()
                ),
                Some(Annotation::new(
                    "Fragment definition target types must be composite types",
                    fragment_definition
                        .type_condition()
                        .named_type()
                        .span()
                        .clone(),
                )),
                Vec::new(),
            ),
            Error::InlineFragmentTargetTypeNotComposite { inline_fragment } => Self::new(
                format!(
                    "`{}` is not a composite type",
                    inline_fragment
                        .type_condition()
                        .map(|tc| tc.named_type().as_ref())
                        .unwrap_or_default()
                ),
                inline_fragment.type_condition().map(|tc| {
                    Annotation::new(
                        "Inline fragment target types must be composite types",
                        tc.named_type().span().clone(),
                    )
                }),
                Vec::new(),
            ),
            Error::FragmentDefinitionUnused {
                fragment_definition,
            } => Self::new(
                format!(
                    "Fragment definition `{}` is unused",
                    fragment_definition.name().as_ref()
                ),
                Some(Annotation::new(
                    "Fragment definition is unused",
                    fragment_definition.name().span().clone(),
                )),
                Vec::new(),
            ),
            Error::FragmentSpreadTargetUndefined { fragment_spread } => Self::new(
                format!(
                    "No fragment defined with name `{}`",
                    fragment_spread.name().as_ref()
                ),
                Some(Annotation::new(
                    "No fragment defined with this name",
                    fragment_spread.name().span().clone(),
                )),
                Vec::new(),
            ),
            Error::FragmentSpreadCycle {
                fragment_definition,
                fragment_spread,
            } => Self::new(
                format!(
                    "Cycle detected in fragment `{}`",
                    fragment_definition.name().as_ref()
                ),
                Some(Annotation::new(
                    "Cycle introduced by fragment spread",
                    fragment_spread.name().span().clone(),
                )),
                vec![Annotation::new(
                    "Affected fragment definition",
                    fragment_definition.name().span().clone(),
                )],
            ),
            Error::FieldSelectionsDoNotMergeDifferingArguments {
                selection_set,
                field_a,
                field_b,
            } => Self::new(
                "Fields in selection set do not merge due to unequal arguments",
                Some(Annotation::new(
                    "Fields in selection set do not merge",
                    selection_set.span().clone(),
                )),
                vec![
                    Annotation::new("First field", field_a.name().span().clone()),
                    Annotation::new("Second field", field_b.name().span().clone()),
                ],
            ),
            Error::FieldSelectionsDoNotMergeDifferingNames {
                selection_set,
                field_a,
                field_b,
            } => Self::new(
                "Fields in selection set do not merge due to unequal field names",
                Some(Annotation::new(
                    "Fields in selection set do not merge",
                    selection_set.span().clone(),
                )),
                vec![
                    Annotation::new("First field", field_a.name().span().clone()),
                    Annotation::new("Second field", field_b.name().span().clone()),
                ],
            ),
            Error::FieldSelectionsDoNotMergeIncompatibleTypes {
                selection_set,
                field_a,
                field_definition_a,
                field_b,
                field_definition_b,
            } => Self::new(
                "Fields in selection set do not merge due to incompatible types",
                Some(Annotation::new(
                    "Fields in selection set do not merge",
                    selection_set.span().clone(),
                )),
                vec![
                    Annotation::new(
                        format!(
                            "First field has type {}",
                            field_definition_a.r#type().as_ref().display_name(),
                        ),
                        field_a.name().span().clone(),
                    ),
                    Annotation::new(
                        format!(
                            "Second field has type {}",
                            field_definition_b.r#type().as_ref().display_name(),
                        ),
                        field_b.name().span().clone(),
                    ),
                ],
            ),
            Error::FragmentSpreadIsNotPossible {
                fragment_spread,
                parent_type,
            } => Self::new(
                format!(
                    "Fragment `{}` cannot be spread for type {}",
                    fragment_spread.name().as_ref(),
                    parent_type.name()
                ),
                Some(Annotation::new(
                    format!("Cannot be spread for type {}", parent_type.name()),
                    fragment_spread.name().span().clone(),
                )),
                Vec::new(),
            ),
            Error::InlineFragmentSpreadIsNotPossible {
                inline_fragment,
                parent_type,
            } => Self::new(
                format!(
                    "Fragment targeting type {} cannot be spread for type {}",
                    inline_fragment
                        .type_condition()
                        .map(|type_condition| type_condition.named_type().as_ref())
                        .unwrap_or_else(|| parent_type.name()),
                    parent_type.name(),
                ),
                Some(Annotation::new(
                    format!("Cannot be spread for type {}", parent_type.name()),
                    inline_fragment.span().clone(),
                )),
                Vec::new(),
            ),
        }
    }
}
