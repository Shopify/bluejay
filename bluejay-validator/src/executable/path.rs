use bluejay_core::executable::ExecutableDocument;
use std::cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd};
use std::hash::{Hash, Hasher};

pub struct Path<'a, E: ExecutableDocument> {
    root: PathRoot<'a, E>,
    members: Vec<&'a E::Selection>,
}

impl<'a, E: ExecutableDocument> Clone for Path<'a, E> {
    fn clone(&self) -> Self {
        Self {
            root: self.root,
            members: self.members.clone(),
        }
    }
}

impl<'a, E: ExecutableDocument> Path<'a, E> {
    pub fn new(root: PathRoot<'a, E>) -> Self {
        Self {
            root,
            members: Vec::new(),
        }
    }

    pub fn root(&self) -> &PathRoot<'a, E> {
        &self.root
    }

    pub fn with_selection(&self, selection: &'a E::Selection) -> Self {
        let mut clone = self.clone();
        clone.members.push(selection);
        clone
    }
}

pub enum PathRoot<'a, E: ExecutableDocument> {
    Operation(&'a E::OperationDefinition),
    Fragment(&'a E::FragmentDefinition),
}

impl<'a, E: ExecutableDocument> Clone for PathRoot<'a, E> {
    fn clone(&self) -> Self {
        match self {
            Self::Operation(o) => Self::Operation(o),
            Self::Fragment(f) => Self::Fragment(f),
        }
    }
}

impl<'a, E: ExecutableDocument> Copy for PathRoot<'a, E> {}

impl<'a, E: ExecutableDocument> Hash for PathRoot<'a, E> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        match self {
            Self::Operation(o) => o.hash(state),
            Self::Fragment(f) => f.hash(state),
        }
    }
}

impl<'a, E: ExecutableDocument> PartialEq for PathRoot<'a, E> {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Operation(l), Self::Operation(r)) if l == r => true,
            (Self::Fragment(l), Self::Fragment(r)) if l == r => true,
            _ => false,
        }
    }
}

impl<'a, E: ExecutableDocument> Eq for PathRoot<'a, E> {}

impl<'a, E: ExecutableDocument> Ord for PathRoot<'a, E> {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (Self::Fragment(l), Self::Fragment(r)) => l.cmp(r),
            (Self::Fragment(_), Self::Operation(_)) => Ordering::Greater,
            (Self::Operation(_), Self::Fragment(_)) => Ordering::Less,
            (Self::Operation(l), Self::Operation(r)) => l.cmp(r),
        }
    }
}

impl<'a, E: ExecutableDocument> PartialOrd for PathRoot<'a, E> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
