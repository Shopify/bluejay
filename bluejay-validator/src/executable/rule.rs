use crate::executable::Visitor;
use bluejay_core::definition::SchemaDefinition;
use bluejay_core::executable::ExecutableDocument;

pub trait Rule<'a, E: ExecutableDocument + 'a, S: SchemaDefinition + 'a>:
    Visitor<'a, E, S> + IntoIterator<Item = Self::Error>
{
    type Error;
}
