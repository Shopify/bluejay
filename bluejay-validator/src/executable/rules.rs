mod argument_names;
mod argument_uniqueness;
mod field_selection_merging;
mod field_selections;
mod fragment_name_uniqueness;
mod fragment_spread_target_defined;
mod fragment_spread_type_exists;
mod fragment_spreads_must_not_form_cycles;
mod fragments_must_be_used;
mod fragments_on_composite_types;
mod leaf_field_selections;
mod lone_anonymous_operation;
mod named_operation_name_uniqueness;
mod operation_type_is_defined;
mod required_arguments;
mod subscription_operation_single_root_field;

pub use argument_names::ArgumentNames;
pub use argument_uniqueness::ArgumentUniqueness;
pub use field_selection_merging::FieldSelectionMerging;
pub use field_selections::FieldSelections;
pub use fragment_name_uniqueness::FragmentNameUniqueness;
pub use fragment_spread_target_defined::FragmentSpreadTargetDefined;
pub use fragment_spread_type_exists::FragmentSpreadTypeExists;
pub use fragment_spreads_must_not_form_cycles::FragmentSpreadsMustNotFormCycles;
pub use fragments_must_be_used::FragmentsMustBeUsed;
pub use fragments_on_composite_types::FragmentsOnCompositeTypes;
pub use leaf_field_selections::LeafFieldSelections;
pub use lone_anonymous_operation::LoneAnonymousOperation;
pub use named_operation_name_uniqueness::NamedOperationNameUniqueness;
pub use operation_type_is_defined::OperationTypeIsDefined;
pub use required_arguments::RequiredArguments;
pub use subscription_operation_single_root_field::SubscriptionOperationSingleRootField;

use crate::executable::{Error, Rule, Visitor};
use bluejay_core::definition::{
    DirectiveLocation, SchemaDefinition, TypeDefinitionReferenceFromAbstract,
};
use bluejay_core::executable::{ExecutableDocument, OperationDefinitionFromExecutableDocument};
use paste::paste;
use std::iter::Chain;

macro_rules! define_rules {
    ( $( $rule:ty ),* $(,)? ) => {
        paste! {
            pub struct Rules<'a, E: ExecutableDocument, S: SchemaDefinition> {
                $([<$rule:snake>]: $rule<'a, E, S>,)*
            }

            impl<'a, E: ExecutableDocument, S: SchemaDefinition> Rule<'a, E, S> for Rules<'a, E, S> {
                fn new(executable_document: &'a E, schema_definition: &'a S) -> Self {
                    Self {
                        $([<$rule:snake>]: $rule::new(executable_document, schema_definition),)*
                    }
                }
            }

            impl<'a, E: ExecutableDocument, S: SchemaDefinition> IntoIterator for Rules<'a, E, S> {
                type Item = Error<'a, E, S>;
                type IntoIter = chain_types!($(<$rule<'a, E, S> as IntoIterator>::IntoIter),*);

                fn into_iter(self) -> Self::IntoIter {
                    chain_iters!($(self.[<$rule:snake>].into_iter()),*)
                }
            }

            impl<'a, E: ExecutableDocument, S: SchemaDefinition> Visitor<'a, E, S> for Rules<'a, E, S> {
                fn visit_operation_definition(&mut self, operation_definition: &'a OperationDefinitionFromExecutableDocument<E>) {
                    $(self.[<$rule:snake>].visit_operation_definition(operation_definition);)*
                }

                fn visit_selection_set(
                    &mut self,
                    selection_set: &'a E::SelectionSet,
                    r#type: &'a TypeDefinitionReferenceFromAbstract<S::TypeDefinitionReference>,
                ) {
                    $(self.[<$rule:snake>].visit_selection_set(selection_set, r#type);)*
                }

                fn visit_field(&mut self, field: &'a E::Field, field_definition: &'a S::FieldDefinition) {
                    $(self.[<$rule:snake>].visit_field(field, field_definition);)*
                }

                fn visit_const_directive(&mut self, directive: &'a E::Directive<true>, location: DirectiveLocation) {
                    $(self.[<$rule:snake>].visit_const_directive(directive, location);)*
                }

                fn visit_variable_directive(&mut self, directive: &'a E::Directive<false>, location: DirectiveLocation) {
                    $(self.[<$rule:snake>].visit_variable_directive(directive, location);)*
                }

                fn visit_fragment_definition(&mut self, fragment_definition: &'a E::FragmentDefinition) {
                    $(self.[<$rule:snake>].visit_fragment_definition(fragment_definition);)*
                }

                fn visit_inline_fragment(&mut self, inline_fragment: &'a E::InlineFragment) {
                    $(self.[<$rule:snake>].visit_inline_fragment(inline_fragment);)*
                }

                fn visit_fragment_spread(&mut self, fragment_spread: &'a E::FragmentSpread) {
                    $(self.[<$rule:snake>].visit_fragment_spread(fragment_spread);)*
                }
            }
        }
    };
}

macro_rules! chain_types {
    ( $first:ty, $( $rest:ty ),+ $(,)? ) => {
        Chain<chain_types!($($rest),+), $first>
    };
    ( $t:ty ) => { $t };
}

macro_rules! chain_iters {
    ( $first:expr, $( $rest:expr ),+ $(,)? ) => {
        chain_iters!($($rest),+).chain($first)
    };
    ( $iter:expr ) => { $iter };
}

define_rules!(
    NamedOperationNameUniqueness,
    LoneAnonymousOperation,
    SubscriptionOperationSingleRootField,
    FieldSelections,
    FieldSelectionMerging,
    OperationTypeIsDefined,
    LeafFieldSelections,
    ArgumentNames,
    ArgumentUniqueness,
    RequiredArguments,
    FragmentNameUniqueness,
    FragmentSpreadTypeExists,
    FragmentsOnCompositeTypes,
    FragmentsMustBeUsed,
    FragmentSpreadTargetDefined,
    FragmentSpreadsMustNotFormCycles,
);
