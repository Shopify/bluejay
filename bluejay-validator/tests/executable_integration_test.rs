use bluejay_parser::{
    ast::definition::{DefinitionDocument, SchemaDefinition},
    ast::executable::ExecutableDocument,
    Error,
};
use bluejay_validator::executable::RulesValidator;

#[test]
fn test_error() {
    with_schema(|schema_definition| {
        insta::glob!("test_data/executable/error/*.graphql", |path| {
            let input = std::fs::read_to_string(path).unwrap();
            let executable_document =
                ExecutableDocument::parse(input.as_str()).expect("Document had parse errors");
            let errors = RulesValidator::validate(&executable_document, &schema_definition);
            let formatted_errors = Error::format_errors(input.as_str(), errors);
            insta::assert_snapshot!(formatted_errors);
        });
    });
}

#[test]
fn test_valid() {
    with_schema(|schema_definition| {
        insta::glob!("test_data/executable/valid/*.graphql", |path| {
            let input = std::fs::read_to_string(path).unwrap();
            let executable_document = ExecutableDocument::parse(input.as_str())
                .expect(format!("Document `{}` had parse errors", path.display()).as_str());
            let errors: Vec<_> = RulesValidator::validate(&executable_document, &schema_definition)
                .into_iter()
                .collect();
            assert!(
                errors.is_empty(),
                "Document `{}` had validation errors",
                path.display()
            )
        });
    });
}

fn with_schema<F: Fn(SchemaDefinition) -> ()>(f: F) {
    let s = std::fs::read_to_string("tests/test_data/executable/schema.graphql").unwrap();
    let definition_document =
        DefinitionDocument::parse(s.as_str()).expect("Schema had parse errors");
    let schema_definition =
        SchemaDefinition::try_from(&definition_document).expect("Schema had errors");
    f(schema_definition)
}
