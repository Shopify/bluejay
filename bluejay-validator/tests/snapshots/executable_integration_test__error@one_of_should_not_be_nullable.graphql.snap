---
source: bluejay-validator/tests/executable_integration_test.rs
expression: formatted_errors
input_file: bluejay-validator/tests/test_data/executable/error/one_of_should_not_be_nullable.graphql
---
Error: Variable $oneOfArg is of type String but must be non-nullable to be used for OneOf Input Object OneOfInput
   ╭─[ one_of_should_not_be_nullable.graphql:2:33 ]
   │
 2 │   oneOf(oneOfArg: { oneOfField: $oneOfArg })
   │                                 ────┬────  
   │                                     ╰────── Variable $oneOfArg is of type String but must be non-nullable to be used for OneOf Input Object OneOfInput
───╯
